### DISCUSSION TITLE
NetCDF File with Rotated Lat/Lon Grid Causing Error

### DISCUSSION THREAD

**Discussion by balbrightwpc** [dtcenter/METplus#2759] [File I/O] [answered]
Hello,

I am trying to run pcp_combine on some experimental West-WRF runs we received from a partner that are in netCDF format. When running the command, I receive the following error:

ERROR  : 
ERROR  : NcCfFile::get_grid_mapping_rotated_latitude_longitude() -> Rotated latitude longitude grid not handled in MET.
ERROR  :

Here are the ncdump details of one of the netCDF files that are being used (I did not include all variables to save space):

ncdump -h west-wrf_gfs_d02_2023-01-09_12_00_00.nc
netcdf west-wrf_gfs_d02_2023-01-09_12_00_00 {
dimensions:
	time = UNLIMITED ; // (1 currently)
	west_east = 717 ;
	south_north = 753 ;
variables:
	double time(time) ;
		time:standard_name = "time" ;
		time:long_name = "Time" ;
		time:units = "hours since 1970-01-01 00:00:00" ;
		time:calendar = "standard" ;
		time:axis = "T" ;
	float lon(south_north, west_east) ;
		lon:standard_name = "longitude" ;
		lon:long_name = "Longitude" ;
		lon:units = "degrees_east" ;
		lon:_CoordinateAxisType = "Lon" ;
	float lat(south_north, west_east) ;
		lat:standard_name = "latitude" ;
		lat:long_name = "Latitude" ;
		lat:units = "degrees_north" ;
		lat:_CoordinateAxisType = "Lat" ;
	float west_east(west_east) ;
		west_east:standard_name = "projection_x_coordinate" ;
		west_east:long_name = "x coordinate of projection" ;
		west_east:units = "degrees_east" ;
		west_east:axis = "X" ;
	float south_north(south_north) ;
		south_north:standard_name = "projection_y_coordinate" ;
		south_north:long_name = "y coordinate of projection" ;
		south_north:units = "degrees_north" ;
		south_north:axis = "Y" ;
	int Rotated_Pole ;
		Rotated_Pole:grid_north_pole_longitude = 180.f ;
		Rotated_Pole:grid_north_pole_latitude = 51.f ;
		Rotated_Pole:grid_mapping_name = "rotated_latitude_longitude" ;
	float precip(time, south_north, west_east) ;
		precip:standard_name = "total_precipitation_amount" ;
		precip:long_name = "Accumulated Total Precipitation Over Simulation" ;
		precip:units = "mm" ;
		precip:grid_mapping = "Rotated_Pole" ;
		precip:coordinates = "lat lon" ;
		precip:_FillValue = 1.e+20f ;
		precip:missing_value = 1.e+20f ;
// global attributes:
		:CDI = "Climate Data Interface version 2.3.0 (https://mpimet.mpg.de/cdi)" ;
		:_NCProperties = "version=2,netcdf=4.7.4,hdf5=1.10.7," ;
		:Conventions = "CF 1.6, Standard Name Table v19" ;
		:source = "/expanse/lustre/scratch/cw3eprod/temp_project/NRT_gfs_backfill/output/2023010900/wrf_2dom/wrfout_d02_2023-01-09_12_00_00" ;
		:institution = "CW3E - Scripps Institution of Oceanography" ;
		:title = "/expanse/lustre/scratch/cw3eprod/temp_project/NRT_gfs_backfill/output/2023010900/cf/wrfcf_gfs_d02_2023-01-09_12_00_00_temp.nc" ;
		:notes = "Created with NCL script:  wrfout_to_cf.ncl v2.0" ;
		:created_by = "Daniel Steinhoff - dcw3eprodf@ucsd.edu" ;
		:creation_date = "Mon Oct  7 19:07:48 PDT 2024" ;
		:history = "Fri Oct 11 11:47:57 2024: cdo -w -L -O merge temp1.nc temp2.nc temp3.nc temp4.nc temp5.nc temp6.nc /expanse/lustre/scratch/cw3eprod/temp_project/NRT_gfs_backfill/output/2023010900/west-wrf_2023010900/west-wrf_gfs_d02_2023-01-09_12_00_00.nc\nFri Oct 11 11:47:34 2024: cdo -w -L -O chname,Z_p,Z500 -setzaxis,surface -sellevel,500 -selname,Z_p wrfcf_gfs_d02_2023-01-09_12_00_00.nc temp1.nc" ;
		:CDO = "Climate Data Operators version 2.3.0 (https://mpimet.mpg.de/cdo)" ;
}

Do you have any advice on how to handle this file? If you need any other information or if you need me to upload one of the files, please let me know. Any help would be appreciated, thank you!

Ben Albright

**Comment by CPKalb** [comment]
Hi Ben:

What version of MET are you using?  Support for rotated lat/lon grids was added in MET version 10.1.1.  So, the first thing to check is to make sure you are running MET version 10.1.1 or higher.  

Assuming you are running 10.1.1 or higher, would you be able to upload a file using the [instructions ](https://github.com/dtcenter/METplus/discussions/954)so we can take a closer look?

Thanks,
Christina

**Reply by balbrightwpc** [reply]
Hi Christina,

Appreciate the response and the information about the versions. I was using an older version and I did have access to 10.1.1 so I tried running my pcp_combine command again. I did not get the error regarding the rotated lat/lon grid. But I did get the following:

ERROR  : 
ERROR  : NcCfFile::get_grid_mapping_rotated_latitude_longitude() -> Didn't find X dimension (degrees_east) in netCDF file.
ERROR  :

I uploaded the two files I am trying to use following your instructions into "albright_data" on the FTP. I would appreciate it if you are able to take a look and see if there's a problem with the netCDF files.

Thanks,
Ben

**Reply by CPKalb** [reply]
Hi Ben:

[Section 4.1.1](https://metplus.readthedocs.io/projects/met/en/latest/Users_Guide/data_io.html#requirements-for-cf-compliant-netcdf) describes what MET needs to read CF compliant netCDF files.  In your files, number 4, or the grid_mapping variable is not present.  However, for latitude and longitude files, MET will try to read them anyway without a grid_mapping variable.  Further down in Section 4.1.1, it lists:

```
When the grid mapping variable is not available, MET detects the latitude_longitude projection in following order:
1. the lat/lon projection from the dimensions
2. the lat/lon projection from the “coordinates” attribute from the data variable
3. the lat/lon projection from the latitude and longitude variables by the “standard_name” attribute
```

I am in the process of getting more information on exactly how those steps work, so we can better understand if there should be a change to the file or if something is not working on our end.  I'll post that information here when I have it.

In the meantime, I was able to get MET to read the file by making a change to the standard_name attribute for the west_east and south_north variables.  If you have access to NCO tools, this is easily accomplished by the following command:

`ncatted -a standard_name,south_north,o,c,grid_latitude -a standard_name,west_east,o,c,grid_longitude west-wrf_gfs_d02_2023-01-09_12_00_00.nc`

With this change, MET read the file.  I am not sure which variable you are using, but I tested with precip.  The data looks right, but I'm not sure it's in the correct location.  

Christina

**Reply by balbrightwpc** [reply]
Hello Christina,

I was able to install NCO tools and ran the command that you suggested. This allowed me to successfully run pcp_combine using the following command:

/opt/MET/METPlus4_1/bin/pcp_combine -subtract west-wrf_gfs_d02_2023-01-10_12_00_00.nc 'name="precip" ; level="(0,*,*)";' west-wrf_gfs_d02_2023-01-09_12_00_00.nc 'name="precip" ; level="(0,*,*)";' test.nc

I really appreciate you taking a look at the files and helping. Feel free to mark the above reply as the answer and if you do find out any more info about the files, please let me know. Thanks again!

Ben

**Reply by JohnHalleyGotway** [reply]
Ben, I pulled the sample files you provided and replicated the same error you're seeing:
```
plot_data_plane west-wrf_gfs_d02_2023-01-10_12_00_00.nc land_mask.ps name="LandMask"; level="(0,*,*)";
WARNING: 
WARNING: NcCfFile::get_grid_mapping_rotated_latitude_longitude() -> Didn't find X dimension (degrees_east) in netCDF file.
WARNING: 
ERROR  : 
ERROR  : NcCfFile::update_grid(Grid &) -> Grid is not ready
ERROR  : 
```
I did find a sample CF-compliant NetCDF file from the UK MetOffice that we use in the MET unit tests:
https://dtcenter.ucar.edu/dfiles/code/METplus/MET/MET_unit_test/unit_test/model_data/nccf/prods_op_ukv_rotated_latlon_20220301_00_000.nc

Here's a sample plot of that data over the UK:
```
plot_data_plane prods_op_ukv_rotated_latlon_20220301_00_000.nc relative_humidity.ps name="relative_humidity"; level="(*,*)";
```
![Screenshot 2024-11-01 at 4 58 40 PM](https://github.com/user-attachments/assets/8020ac79-0265-44b2-86d3-3801f3d7f071)

Comparing the `ncdump -h` output, my initial guess is that the problem is that you're lat/lon variables are 2-dimensional:
```
float lon(south_north, west_east) ;
float lat(south_north, west_east) ;
```
Where the MetOffice one that's working has 1-dimensional lat/lon variables:
```
float grid_latitude(grid_latitude) ;
double grid_longitude(grid_longitude) ;
```
I wonder if we reduced your `lat` and `lon` variables to 1-D if we'd get a better result?

**Reply by CPKalb** [reply]
There are 1d lat/lon variables in the file.  In this case, they are called west_east and north_south.  I was originally confused since the long names say projection_x_coordinate and projection_y_coordinate.  But, looking at the values, I think they are latitude and longitude.

**Reply by JohnHalleyGotway** [reply]
I was able to make some more progress... but issues still remain.

Based on [this line](https://github.com/dtcenter/MET/blob/c5a2155ada4773b2d3c7f50d5dea8f1dff791153/src/libcode/vx_data2d_nccf/nccf_file.cc#L2622) of `nccf_file.cc`, I see that we're looking for NetCDF variables whose `standard_name` attribute is set to `grid_latitude` and `grid_longitude` when reading the grid information. So I used `ncatted` to modify the standard_name attributes of the west_east and south_north variables in the sample file:
```
ncatted \
-a standard_name,west_east,o,c,"grid_longitude" \
-a standard_name,south_north,o,c,"grid_latitude" \
west-wrf_gfs_d02_2023-01-09_12_00_00.nc \
west-wrf_gfs_d02_2023-01-09_12_00_00_MOD.nc
```
Plotting that does run, but the plot includes no background map data, which usually means that it's not in the correct spot on earth:
```
plot_data_plane west-wrf_gfs_d02_2023-01-09_12_00_00_MOD.nc LandMask.ps 'name="LandMask"; level="(0,*,*)";' -v 4
DEBUG 4: Grid information:
DEBUG 4:    Projection: Rotated Lat/Lon
DEBUG 4:    Nx: 717
DEBUG 4:    Ny: 753
DEBUG 4:    rot_lat_ll: 30.343
DEBUG 4:    rot_lon_ll: 133.155
DEBUG 4:    delta_rot_lat: 0.027
DEBUG 4:    delta_rot_lon: 0.030
DEBUG 4:    wrapLon: false
DEBUG 4:    true_lat_south_pole: -51.000
DEBUG 4:    true_lon_south_pole: 0.000
DEBUG 4:    aux_rotation: 0.000
```
Here's the image:
<img width="371" alt="Screenshot 2024-11-04 at 9 39 42 AM" src="https://github.com/user-attachments/assets/c35bfc8f-0b4c-4e0a-8aef-a78e4c5ea2b5">
Then I ran regrid_data_plane to regrid this to a global grid (NCEP Grid 4) and plotted it to see where MET thinks this data lives:
```
regrid_data_plane west-wrf_gfs_d02_2023-01-09_12_00_00_MOD.nc G004 LandMask_G004.nc  -field 'name="LandMask"; level="(0,*,*)";'
plot_data_plane LandMask_G004.nc LandMask_G004.ps 'name="LandMask"; level="(*,*)";'
```
<img width="457" alt="Screenshot 2024-11-04 at 9 41 17 AM" src="https://github.com/user-attachments/assets/d1c5c76d-a233-44b1-9819-6d4eeed256f4">

So data the should clearly be over the western United States is in the middle of the Pacific Ocean. Clearly, there's a problem. However I don't know exactly what it is. Is it a problem with the projection information itself or is it a bug in MET's processing of that grid information?

**Reply by balbrightwpc** [reply]
Hi John,

Appreciate the reply and you continuing to look into the issue. I ran into the same issue when trying to plot the data as well after renaming the data. I believe one of my co-workers was able to plot the raw data correctly. I have asked him to either post here or relay to me what he did so stay tuned.

**Reply by tomerburg** [reply]
Hi all,

The metadata in these files are different than files I normally work with - the 1D coordinates "east_west" and "south_north" are technically in degrees longitude and latitude, respectively, but don't actually correspond to the coordinates of the data. That would be the "lon" and "lat" 2D arrays. When I plotted it with Python/Cartopy, using the "lon" and "lat" arrays as the x and y coordinates, I was able to get the data correctly plotted:

![image](https://github.com/user-attachments/assets/36ee7dc4-2052-4fa5-8fd2-a975b894ce21)

I don't have enough MET expertise to know whether it's a bug on MET's side or with how the data is provided into MET, but hopefully this is enough information to guide you in the right direction.

**Reply by CPKalb** [reply]
@JohnHalleyGotway, correct me if I'm wrong, but I think MET needs the latitude and longitude for rotated lat/lon grids to be 1d variables.  However, from the previous post, those variables don't correspond to the actual data in the file.  I ran some testing on this.  Specifically, I edited the netCDF file so that the dimensions are lat and lon (rather than west_east and south_north).  I also edited the lat and lon variables to have standard name attributes of grid_latitude and grid_longitude.  However, the file still does not go through MET.  I get the following error:

```
ERROR  : 
ERROR  : NcCfFile::get_grid_mapping_rotated_latitude_longitude() -> Coordinate variables don't match dimension sizes in netCDF file.
ERROR  : 
```

I'm guessing you will need to read this file in using a python embedding script, as I don't think it can be read into MET in the format it is currently in.

Tina

**Reply by JohnHalleyGotway** [reply, marked as answer]
Thanks for the discussion on this. I can confirm that MET really is using the data lat/lon information from the 1D "east_west" and "south_north" coordinate variables to geo-locate this data. It's does not use the 2D lat/lon variables at all. And I believe that MET really is processing this data as it should based on the CF-conventions for [Rotated Pole](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.12/cf-conventions.html#_rotated_pole) data. I certainly don't see any obvious bugs in the MET code that warrant a bugfix.

There are some CF-compliance checker utilities that you could try out:
https://cfconventions.org/software.html

They might be useful in figuring out how to specify the grid information in a cf-compliant way.

**Reply by balbrightwpc** [reply]
I appreciate you both taking the time to look into this issue. We are back in contact with our partners that provided the data and are working on a solution to format them differently.

**Reply by CPKalb** [reply]
Thanks for the update.  I'm going to go ahead and close this issue then.  Please feel free to start up a new issue if you encounter problems with the newly formatted data files or other MET/METplus issues.

Christina
